src/hotspot/share/gc/g1/g1ConcurrentMark.cpp

```diff
@@ -1131,13 +1131,12 @@ class G1UpdateRemSetTrackingBeforeRebuildTask : public WorkerTask {
     // Distribute the given marked bytes across the humongous object starting
     // with hr and note end of marking for these regions.
     void distribute_marked_bytes(HeapRegion* hr, size_t marked_bytes) {
-      size_t const obj_size_in_words = cast_to_oop(hr->bottom())->size();
-
       // "Distributing" zero words means that we only note end of marking for these
-      // regions.
-      assert(marked_bytes == 0 || obj_size_in_words * HeapWordSize == marked_bytes,
+      // regions; also, we should not access their header any more them as their
+      // klass may have been unloaded.
+      assert(marked_bytes == 0 || cast_to_oop(hr->bottom())->size() * HeapWordSize == marked_bytes,
```

Comment:
Is it possible to separate these two cases into two methods, for instance? Taking a step back, why do we even need to call `note_end_of_marking` on these effectively empty regions?

Reply:
We need to call `HeapRegion::note_end_of_marking` on them, even if they are empty for completeness. It's not completely necessary because reclamation will probably reset them correctly, but it's easier to reason if they (empty and nonempty regions) are handled the same to me.

I.e. so _all_ regions get `note_start/end_of_marking` called.

Reply:
I do not think it is easier to understand to separate these two cases (empty/non-empty) regions: both distributing bytes (it's fine to distribute 0 bytes) and being consistent with calling the start/end notifications (and `note_end` works as expected on empty regions too) for all regions is easier to understand to me as there are no exceptions.

Even if they do not do anything "meaningful" other than resetting some internal state that is later overwritten for these specially handled empty regions anyway.

Reply:
> so all regions have note_start/end_of_marking called.

Currently, the pair is invoked on all regions and the filtering (skipping young-region for example) is done inside these methods. However, the logic why a particular kind of region can (should) be skipped really belongs to the caller. The region itself doesn't know how to react to marking-start/end. (This is kind of tied to the ticket of moving marking-related fields outside region.)

> why have that exception?

Because live-region and effective-region are diff, and mixing them causes confusion. I think the existence of the new comment "we should not access their header any more them..." demonstrates that it's not super obvious why the current code (before this PR) is problematic.

Reply:
> > so all regions have note_start/end_of_marking called.
> 
> Currently, the pair is invoked on all regions and the filtering (skipping young-region for example) is done inside these methods. However, the logic why a particular kind of region can (should) be skipped really belongs to the caller. The region itself doesn't know how to react to marking-start/end. (This is kind of tied to the ticket of moving marking-related fields outside region.)

We already agreed that these `note*` methods are basically part of the caller, placed in the wrong location because the members it accesses are in the wrong location. I do not think messing with this here in this CR half-heartedly is a good idea. As soon as the work to move TAMS and PB starts, this is going to change anyway and is imho a more appropriate time to reconsider this (and will probably naturally fix itself).

The problematic code is assertion code, which quite often accesses internals one normally would not. The regular code is independent of whether the region's klass is live or not after all.

Deleting this assert would fix the issue at hand as well. Another option would be to just not do class unloading this early; there is no particular reason to do it right after marking completed.

> 
> > why have that exception?
> 
> Because live-region and effective-region are diff, and mixing them causes confusion.

What is an "effective-region" in this context? I do not understand this sentence.

>I think the existence of the new comment "we should not access their header any more them..." demonstrates that it's not super obvious why the current code (before this PR) is problematic.

To me this change indicates that the sanity check code (this problematic statement is part of sanity check code - regular code does not use it) is doing things it should not. The original author (me I guess) correctly considered that we already unloaded classes super-early for some reason, wanted to have some extra check there, but then botched the refactoring (factoring out the `obj_size_in_words` calculation for the two uses).
That particular new comment is only to make it abundantly clear to not factor out any kind of `obj_size` calculation any more (I removed the second use in this change).
Maybe not adding the comment would have been better, as the `marked_bytes == 0` predicate already indicates just that (and the second use of `obj_size_in_words` is gone)

Looking at the code again, another source for confusion is maybe wrong comment placement. I will improve these.